edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/CrdWatcher.cs

```diff
@@ -0,0 +1,812 @@
+                    return;
+                }
+
+                // only operate on the device that matches this operator.
+                if (String.Equals(customObject.Metadata.Name, this.resourceName, StringComparison.OrdinalIgnoreCase))
+                {
+                    V1ServiceList currentServices = await this.client.ListNamespacedServiceAsync(KubeUtils.K8sNamespace, labelSelector: this.deploymentSelector);
+                    V1DeploymentList currentDeployments = await this.client.ListNamespacedDeploymentAsync(KubeUtils.K8sNamespace, labelSelector: this.deploymentSelector);
+                    this.logger.DeploymentStatus(type, this.resourceName);
+                    switch (type)
+                    {
+                        case WatchEventType.Added:
+                        case WatchEventType.Modified:
+                            this.ManageDeployments(currentServices, currentDeployments, customObject);
+                            break;
+
+                        case WatchEventType.Deleted:
+                            {
+                                // Delete the deployment.
+                                // Delete any services.
+                                IEnumerable<Task<V1Status>> removeServiceTasks = currentServices.Items.Select(i => this.client.DeleteNamespacedServiceAsync(i.Metadata.Name, KubeUtils.K8sNamespace, new V1DeleteOptions()));
+                                await Task.WhenAll(removeServiceTasks);
+                                IEnumerable<Task<V1Status>> removeDeploymentTasks = currentDeployments.Items.Select(
+                                    d => this.client.DeleteNamespacedDeployment1Async(
+                                        d.Metadata.Name,
+                                        KubeUtils.K8sNamespace,
+                                        new V1DeleteOptions(propagationPolicy: "Foreground"),
+                                        propagationPolicy: "Foreground"));
+                                await Task.WhenAll(removeDeploymentTasks);
+                            }
+                            break;
+                        case WatchEventType.Error:
+                            this.logger.DeploymentError();
+                            break;
+                    }
+                }
+                else
+                {
+                    this.logger.DeploymentNameMismatch(customObject.Metadata.Name, this.resourceName);
+                }
+            }
+        }
+
+        private string DeploymentName(string moduleId) => KubeUtils.SanitizeK8sValue(moduleId);
+
+        private async void ManageDeployments(V1ServiceList currentServices, V1DeploymentList currentDeployments, EdgeDeploymentDefinition<TConfig> customObject)
+        {
+            var desiredModules = ModuleSet.Create(customObject.Spec.ToArray());
+            var moduleIdentities = await this.moduleIdentityLifecycleManager.GetModuleIdentitiesAsync(desiredModules, this.currentModules);
```

Comment:
>GetModuleIdentitiesAsync [](start = 77, length = 24)

The default behavior for this is to only return new identities between current and desired.  This might cause you some trouble during integration.

Reply:
That problem was found late in release process. We actually made a new version of the lifecycle manager in the Kubernetes project which overrode that behavior, but it would probably make more sense to create a different method on the `IModuleIdentityLifecycleManager` interface.

---
In reply to: [303540007](https://github.com/Azure/iotedge/pull/1437#discussion_r303540007) [](ancestors = 303540007)